# 📊 Introduction to Exploratory Data Analysis (EDA)

## 🎯 Learning Objectives

After completing this notebook, you will be able to:

- Understand what Exploratory Data Analysis (EDA) is.
- Learn why EDA is important before building ML models.
- Understand the EDA workflow.
- Explore the Titanic dataset using Pandas.
- Identify different types of variables.

## 📖 Introduction

Exploratory Data Analysis (EDA) is the process of examining, summarizing, and visualizing data before applying machine learning algorithms.

The main goal of EDA is to understand the dataset, identify patterns, detect anomalies, and prepare the data for further analysis.

## ❓ What is Exploratory Data Analysis?

EDA is the first analytical step after data collection and cleaning.

It helps answer questions such as:

- What does the dataset contain?
- Are there missing values?
- Are there duplicate records?
- What are the distributions of features?
- Are there relationships between variables?
- Are there outliers?

## 💡 Why Perform EDA?

EDA helps to:

- Understand the structure of the dataset.
- Detect missing values and duplicates.
- Identify outliers.
- Discover trends and patterns.
- Select useful features.
- Improve machine learning model performance.

## 🔄 Typical EDA Workflow

1. Load the dataset.
2. Understand the data.
3. Inspect data types.
4. Check missing values.
5. Analyze numerical features.
6. Analyze categorical features.
7. Identify relationships between variables.
8. Visualize the data.
9. Draw conclusions.

In [1]:
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv(r"D:\Sanskruti\sanskruti document\Machine-Learning-From-Scratch\datasets\train.csv")

In [3]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [4]:
df.shape

(891, 12)

In [5]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [6]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [8]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [9]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [10]:
df.duplicated().sum()

np.int64(0)

In [11]:
df.select_dtypes(include=["int64","float64"]).columns

Index(['PassengerId', 'Survived', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare'], dtype='object')

In [12]:
df.select_dtypes(include=["object"]).columns

Index(['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked'], dtype='object')

## 🔍 Initial Observations

- The dataset contains 891 passenger records.
- Several columns contain missing values.
- Both numerical and categorical features are present.
- The Titanic dataset is suitable for classification problems.

## ✅ Best Practices

- Always inspect the dataset before analysis.
- Understand feature meanings.
- Check data types.
- Identify missing values and duplicates.
- Never start model building without EDA.

## 🌍 Real-World Scenario

A hospital wants to predict whether a patient is at risk of heart disease. Before building the prediction model, data scientists first perform EDA to understand patient demographics, identify missing medical records, detect unusual values, and study relationships between health indicators. This helps build more reliable and accurate predictive models.

## 📝 Summary

In this notebook, we introduced Exploratory Data Analysis (EDA), understood its importance, explored the Titanic dataset, inspected its structure, identified missing values, examined data types, and reviewed basic statistical summaries. EDA forms the foundation for every successful data analysis and machine learning project.